iml-agent/src/action_plugins/lustre/snapshot/parse.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use chrono::offset::Utc;
+use chrono::DateTime;
+use combine::{
+    attempt, choice, eof,
+    error::{ParseError, StreamError},
+    many1,
+    parser::{char, repeat::take_until},
+    satisfy, skip_many,
+    stream::{Stream, StreamErrorFor},
+    Parser,
+};
+use iml_wire_types::snapshot::{Detail, Snapshot, Status};
+use std::collections::HashMap;
+
+#[derive(PartialEq, Debug)]
+enum Segment {
+    Snapshot(Snapshot),
+    Detail(Detail),
+}
+
+fn spaces<I>() -> impl Parser<I, Output = ()>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    skip_many(satisfy(|c| c == ' ' || c == '\t'))
+}
+
+fn eol<I>() -> impl Parser<I, Output = ()>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    spaces().skip(choice((char::newline().map(|_| ()), eof())))
+}
+
+fn whitespace<I>() -> impl Parser<I, Output = ()>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    skip_many(combine::parser::char::space())
+}
+
+fn delimiter<I>() -> impl Parser<I, Output = ()>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    spaces().skip(char::char(':').skip(spaces()))
+}
+
+fn entry<I>() -> impl Parser<I, Output = (String, String)>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (
+        spaces()
+            .with(take_until(attempt(delimiter().or(eol()))))
+            .skip(delimiter()),
+        take_until(attempt(eol())).skip(eol()),
+    )
+}
+
+fn parse_date(mut s: String) -> Result<DateTime<Utc>, &'static str> {
+    // Can't parse without a timezone.
+    // lctl does not print timezone, but uses gettimeofday()
+    // to get the time and ctime() to print it.
+    s.push_str(" +00:00");
+    DateTime::parse_from_str(&s, "%a %b %e %T %Y %#z")
+        .map(|t| t.with_timezone(&Utc))
+        .map_err(|_| "date string like 'Fri Aug  7 16:43:06 2020'")
+}
+
+fn mk_detail(mut m: HashMap<String, String>) -> Result<Detail, &'static str> {
+    let fsname = m.remove("snapshot_fsname").ok_or("snapshot_fsname")?;
+    let modify_time_s = m.remove("modify_time").ok_or("modify_time")?;
+    let create_time_s = m.remove("create_time").ok_or("create_time")?;
```

Comment:
Why return the field name over returning an Error?

Reply:
Actually, it returns a message for `expected_static_message()`. I am ok with improving it, but for the beginning simple static string looked fine.